app/services/doi_service.rb

```diff
@@ -2,83 +2,149 @@
 
 require 'data_cite'
 
-class DoiService
+module DoiService
   class Error < StandardError; end
 
   def self.call(resource, client_source: nil, metadata_source: nil)
-    instance = new(resource)
+    strategy_class = [WorkAndVersionStrategy, CollectionStrategy].find { |klass| klass.applicable_to?(resource) }
+
+    raise ArgumentError, "DoiService cannot be called with a #{resource.class.name}" if strategy_class.blank?
+    raise Error.new("Cannot mint a doi for an invalid resource: #{resource.inspect}") unless resource.valid?
+
+    instance = strategy_class.new(resource)
     instance.client_source = client_source
     instance.metadata_source = metadata_source
     instance.call
   end
 
-  attr_reader :resource
+  def self.publish_doi(doi:, metadata_adapter:, client:)
+    metadata = metadata_adapter
+      .tap(&:validate!)
+      .attributes
 
-  attr_writer :client_source,
-              :metadata_source
+    response_doi, _response_metadata = client.publish(
+      doi: doi,
+      metadata: metadata
+    )
 
-  def initialize(resource)
-    @resource = resource
+    response_doi
+  end
 
-    raise Error.new("Cannot mint a doi for an invalid resource: #{resource.inspect}") unless resource.valid?
+  class WorkAndVersionStrategy
+    attr_reader :resource,
+                :work_version
+
+    attr_writer :client_source,
+                :metadata_source
 
-    if resource.is_a? WorkVersion
-      @work_version = resource
-    elsif resource.is_a? Work
-      @work_version = resource.latest_version
-    else
-      raise ArgumentError, 'DoiService expects resource to be a Work or WorkVersion'
+    def self.applicable_to?(obj)
+      obj.is_a?(Work) || obj.is_a?(WorkVersion)
     end
-  end
 
-  def call
-    if resource.doi.blank? && work_version.draft?
-      register_and_save_new_doi
-    elsif resource.doi.present? && work_version.draft?
-      # No-op
-    else
-      existing_doi = resource.doi.presence
-      should_save_new_doi = existing_doi.blank?
-
-      publish_doi(
-        doi: existing_doi, # will register a new doi if nil
-        update_resource: should_save_new_doi
-      )
+    def initialize(resource)
+      unless self.class.applicable_to?(resource)
+        raise ArgumentError, "#{self.class.name} expects resource to be a Work or WorkVersion"
```

Comment:
This line isn't getting covered in the tests, but that might not be important.

Reply:
It's true that it's not being covered in tests, because `DoiService.call` will never instantiate one of these instances unless it has already passed the `applicable_to?` test when selecting a strategy. This line probably doesn't even need too be there but I thought it was worth being prudent. ¯\_(ツ)_/¯